src/main/kotlin/com/exactpro/th2/conn/dirty/tcp/core/Pipe.kt

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2022-2022 Exactpro (Exactpro Systems Limited)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.exactpro.th2.conn.dirty.tcp.core
+
+import mu.KotlinLogging
+import org.jctools.queues.MpscChunkedArrayQueue
+import java.util.Queue
+import java.util.concurrent.Executor
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.function.Consumer
+
+class Pipe<E>(
+    private val name: String,
+    private val queue: Queue<E>,
+    private val executor: Executor,
+    private val consumer: Consumer<E>,
+) : Runnable, AutoCloseable {
+    private val size = AtomicInteger()
+
+    @Volatile var isOpen = true
+        private set
+
+    fun send(item: E, timeout: Long = 5000): Boolean {
+        if (offer(item)) return true
+
+        LOGGER.warn("Pipe is full: $name")
+
+        val deadline = System.currentTimeMillis() + timeout
+
+        while (deadline > System.currentTimeMillis()) {
+            if (offer(item)) return true
+            Thread.sleep(1)
+        }
+
+        return false
+    }
+
+    private fun offer(item: E): Boolean {
+        check(isOpen) { "Pipe is closed: $name" }
+        if (!queue.offer(item)) return false
+        if (size.incrementAndGet() == 1) schedule()
+        return true
+    }
+
+    private fun schedule() {
+        try {
+            executor.execute(this)
+        } catch (e: Exception) {
+            isOpen = false
+            LOGGER.error(e) { "Failed to schedule consume task for pipe: $name" }
+        }
+    }
+
+    override fun run() {
+        try {
+            consumer.accept(queue.poll())
+        } catch (e: Error) {
+            throw e
+        } catch (e: InterruptedException) {
+            Thread.currentThread().interrupt()
+        } catch (t: Throwable) {
+            LOGGER.error(t) { "Failed to consume element from pipe: $name" }
+        } finally {
+            if (isOpen && size.decrementAndGet() > 0) schedule()
+        }
+    }
+
+    override fun close() {
+        isOpen = false
+    }
+
+    companion object {
+        private val LOGGER = KotlinLogging.logger {}
+
+        fun <E> Executor.newPipe(
+            name: String,
+            queue: Queue<E> = MpscChunkedArrayQueue(1024),
```

Comment:
Our conn should be able to take a load (min level is 10k). Fixed value of max capacity can be bottle neck. Maybe add an option to the custom configuration.

Reply:
> Maybe add an option to the custom configuration.

I doubt that an average user would ever touch it. Same goes for thread count - we're considering on removing it in the future versions)

Reply:
Let's conduct load test to choose suitable value